commit 9a83564c58b7f6363141a8f1d0c87c89a5ebab5d
Author: Peter Eisentraut <peter_e@gmx.net>
Date:   Sat Mar 19 11:03:22 2016 +0100

    Allow SSL server key file to have group read access if owned by root
    
    We used to require the server key file to have permissions 0600 or less
    for best security.  But some systems (such as Debian) have certificate
    and key files managed by the operating system that can be shared with
    other services.  In those cases, the "postgres" user is made a member of
    a special group that has access to those files, and the server key file
    has permissions 0640.  To accommodate that kind of setup, also allow the
    key file to have permissions 0640 but only if owned by root.
    
    From: Christoph Berg <myon@debian.org>
    Reviewed-by: Alvaro Herrera <alvherre@alvh.no-ip.org>

--- a/src/backend/libpq/be-secure.c
+++ b/src/backend/libpq/be-secure.c
@@ -770,8 +770,30 @@ initialize_SSL(void)
 					 errmsg("could not access private key file \"%s\": %m",
 							SERVER_PRIVATE_KEY_FILE)));
 
+		if (!S_ISREG(buf.st_mode))
+			ereport(FATAL,
+					(errcode(ERRCODE_CONFIG_FILE_ERROR),
+					 errmsg("private key file \"%s\" is not a regular file",
+							SERVER_PRIVATE_KEY_FILE)));
+
+		/*
+		 * Refuse to load files owned by users other than us or root.
+		 *
+		 * XXX surely we can check this on Windows somehow, too.
+		 */
+#if !defined(WIN32) && !defined(__CYGWIN__)
+		if (buf.st_uid != geteuid() && buf.st_uid != 0)
+			ereport(FATAL,
+					(errcode(ERRCODE_CONFIG_FILE_ERROR),
+					 errmsg("private key file \"%s\" must be owned by the database user or root",
+							SERVER_PRIVATE_KEY_FILE)));
+#endif
+
 		/*
-		 * Require no public access to key file.
+		 * Require no public access to key file. If the file is owned by us,
+		 * require mode 0600 or less. If owned by root, require 0640 or less
+		 * to allow read access through our gid, or a supplementary gid that
+		 * allows to read system-wide certificates.
 		 *
 		 * XXX temporarily suppress check when on Windows, because there may
 		 * not be proper support for Unix-y file permissions.  Need to think
@@ -779,12 +801,13 @@ initialize_SSL(void)
 		 * directory permission check in postmaster.c)
 		 */
 #if !defined(WIN32) && !defined(__CYGWIN__)
-		if (!S_ISREG(buf.st_mode) || buf.st_mode & (S_IRWXG | S_IRWXO))
+		if ((buf.st_uid == geteuid() && buf.st_mode & (S_IRWXG | S_IRWXO)) ||
+			(buf.st_uid == 0 && buf.st_mode & (S_IWGRP | S_IXGRP | S_IRWXO)))
 			ereport(FATAL,
 					(errcode(ERRCODE_CONFIG_FILE_ERROR),
-				  errmsg("private key file \"%s\" has group or world access",
-						 SERVER_PRIVATE_KEY_FILE),
-				   errdetail("Permissions should be u=rw (0600) or less.")));
+					 errmsg("private key file \"%s\" has group or world access",
+							SERVER_PRIVATE_KEY_FILE),
+					 errdetail("File must have permissions u=rw (0600) or less if owned by the database user, or permissions u=rw,g=r (0640) or less if owned by root.")));
 #endif
 
 		if (SSL_CTX_use_PrivateKey_file(SSL_context,
